cabal-install/src/Distribution/Client/ProjectPlanning.hs

```diff
@@ -2862,26 +2869,105 @@ setRootTargets targetAction perPkgTargetsMap =
 --
 pruneInstallPlanPass1 :: [ElaboratedPlanPackage]
                       -> [ElaboratedPlanPackage]
-pruneInstallPlanPass1 pkgs =
-    map (mapConfiguredPackage fromPrunedPackage)
-        (fromMaybe [] $ Graph.closure graph roots)
+pruneInstallPlanPass1 pkgs
+    -- if there are repl targets, we need to do a bit more work
+    | anyReplTarget = final_final_graph
+
+    -- otherwise we'll do less
+    | otherwise     = pruned_packages
   where
+    pkgs' :: [InstallPlan.GenericPlanPackage IPI.InstalledPackageInfo PrunedPackage]
     pkgs' = map (mapConfiguredPackage prune) pkgs
-    graph = Graph.fromDistinctList pkgs'
-    roots = mapMaybe find_root pkgs'
 
+    prune :: ElaboratedConfiguredPackage -> PrunedPackage
     prune elab = PrunedPackage elab' (pruneOptionalDependencies elab')
       where elab' =
                 setDocumentation
               $ addOptionalStanzas elab
 
+    graph = Graph.fromDistinctList pkgs'
+
+    roots :: [UnitId]
+    roots = mapMaybe find_root pkgs'
+
+    -- Make a closed graph by calculating the closure from the roots
+    pruned_packages :: [ElaboratedPlanPackage]
+    pruned_packages =  map (mapConfiguredPackage fromPrunedPackage) (fromMaybe [] $ Graph.closure graph roots)
+
+    closed_graph :: Graph.Graph ElaboratedPlanPackage
+    closed_graph = Graph.fromDistinctList pruned_packages
+
+    -- whether any package has repl targets enabled.
+    anyReplTarget :: Bool
+    anyReplTarget = any is_repl_gpp pkgs' where
+      is_repl_gpp (InstallPlan.Configured pkg) = is_repl_pp pkg
+      is_repl_gpp _                            = False
+
+      is_repl_pp (PrunedPackage elab _) = not (null (elabReplTarget elab))
+
```

Comment:
I would appreciate it if we had a high-level description of what we are doing here. Reading the code is possible, but just takes a long time, and kind of requires us to already know where we are going.

Taking a stab at an explanation, feel free to discuss and discard:
```suggestion

    -- Note [Pruning for MultiRepl]:
    -- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    --
    -- For a multi-repl session, where we load more than target into a GHCi repl,
    -- it is required to uphold the so-called *closure property*.
    -- This property, whose exact Note you can read in the GHC codebase, states
    -- roughly:
    --
    -- * if a component you want to load into a repl session transitively depends on a
    --   component which transitively depends on another component you want to
    --   load into the repl, then this component needs to be loaded
    --   into the repl session as well.
    --
    -- We make sure here, that this property is upheld, by calculating the
    -- graph of components that we need to load into the repl given 'roots'.
    --
    -- Practically, this is simply achieved by traversing all dependencies of
    -- our roots (graph closure), and then from this closed graph, we calculate
    -- the reverse closure, which gives us all components that depend on
    -- 'roots'. Thus, the result is a list of components that we need to load
    -- into the repl to uphold the closure property.

```